File: 567-permutation-in-string/permutation-in-string.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {boolean}
 */
var checkInclusion = function (s1, s2) {
  const s1Chars = new Array(26).fill(0);
  const s2Chars = new Array(26).fill(0);

  for (let i = 0; i < s1.length; i++) {
    s1Chars[s1.charCodeAt(i) - "a".charCodeAt(0)]++;
  }

  let left = 0;
  let right = left + s1.length;
  while (right < s2.length + 1) {
    s2Chars.fill(0);
    for (let i = left; i < right; i++) {
      s2Chars[s2.charCodeAt(i) - "a".charCodeAt(0)]++;
    }

    let matches = true;
    for (let i = 0; i < 26; i++) {
      if (s1Chars[i] != s2Chars[i]) {
        matches = false;
      }
    }
    if (matches) return true;

    left++;
    right++;
  }
  return false;
};

File: 567-permutation-in-string/permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
    int[] s1Chars = new int[26];
    int[] s2Chars = new int[26];

    for (int i = 0; i < s1.length(); i++) {
      s1Chars[s1.charAt(i) - 'a']++;
    }

    int left = 0;
    int right = left + s1.length();
    while (right < s2.length() + 1) {
      Arrays.fill(s2Chars, 0);
      for (int i = left; i < right; i++) {
        s2Chars[s2.charAt(i) - 'a']++;
      }

      boolean matches = true;
      for (int j = 0; j < 26; j++) {
        if (s1Chars[j] != s2Chars[j]) {
          matches = false;
        }
      }
      if (matches) {
        return true;
      }

      left++;
      right++;
    }
    return false;
    }
}